### PR TITLE
[WIP] Allow players to win the game when someone's fleet is destroyed

### PR DESCRIPTION
**Pull request by benjaminpayoux**
[Trello](https://trello.com/c/rdsVYDzO/5-05-as-peter-i-would-like-the-game-to-tell-me-if-i-won-or-not)

# TODO
- [X] Display players fleet state
- [X] End the game when a player has hit all ships 
- [X] Tests

# Result
last

### CODE REVIEW

File: src/Game.py [hunk 1]
```diff
@@ -19,10 +19,12 @@ def initGame(config1, config2, size):
         PLAYER_1: {
             "primary": getBoardFromConfig(player1Config),
             "opponent": initBoard(10),
+            "life": FLEET_LIFE
```
Comment (zyhou) [inline src/Game.py:22]: Interesting to offer the posibility to make different life number for each player

File: src/Game.py [hunk 2]
```diff
@@ -33,3 +35,17 @@ def initGame(config1, config2, size):
         currentPlayer = PLAYER_2
 
     return boards, currentPlayer
+
+def decrementTargetFleetLife(boards, attacker):
+    boardsCopy = deepcopy(boards)
+
+    if attacker == PLAYER_1:
+        boardsCopy[PLAYER_2]["life"] -= 1
```
Comment (Luwangel) [inline src/Game.py:43]: You can replace PLAYER_2 by attacker

``` py
boardsCopy[attacker]["life"] -= 1
```

File: src/Game.py [hunk 3]
```diff
@@ -33,3 +35,17 @@ def initGame(config1, config2, size):
...
+    else:
+        boardsCopy[PLAYER_1]["life"] -= 1
+
+    return boardsCopy
+
+def gameIsWon(boards):
+    return boards[PLAYER_1]["life"] <= 0 or boards[PLAYER_2]["life"] <= 0
```
Comment (zyhou) [inline src/Game.py:50]: image

File: src/Game.py [hunk 4]
```diff
@@ -19,10 +19,12 @@ def initGame(config1, config2, size):
         PLAYER_1: {
             "primary": getBoardFromConfig(player1Config),
```
Comment (Luwangel) [inline src/Game.py:20]: The key "primary" could be renamed to "fleet" or "board" (or another name)

File: src/ui.py [hunk 5]
```diff
@@ -12,22 +12,100 @@ def displayPlayerBoard(boards, player, hit):
...
     clear()
 
     if (currentPlayer == PLAYER_1):
+        displayFleetLife(PLAYER_1, boards)
         displayBoard(boards[PLAYER_1]["opponent"])
     else:
+        displayFleetLife(PLAYER_2, boards)
         displayBoard(boards[PLAYER_2]["opponent"])
 
     if hit:
```
Comment (zyhou) [inline src/ui.py:68]: two condition `if`

**Review by zyhou — CHANGES_REQUESTED** [review]

File: src/ui.py [hunk 6]
```diff
@@ -12,22 +12,100 @@ def displayPlayerBoard(boards, player, hit):
...
+
+def getPlayerName(player):
+    name = 'PLAYER'
+
+    if player == PLAYER_1:
+        name = 'PLAYER 1'
+    else:
+        name = 'PLAYER 2'
+
+    return name
```
Comment (Luwangel) [inline src/ui.py:88]:
```suggestion
def getPlayerName(player):
    if player == PLAYER_1:
        return 'PLAYER 1'
    else:
        return 'PLAYER 2'
```

You don't need a variable.

**Review by Luwangel — CHANGES_REQUESTED** [review]

File: src/ui.py (hunk 6 above)
Comment (benjaminpayoux) [reply src/ui.py:88]: I prefer to have only one return at the end of a function. That's why it is more explicit. What do you think ?

Comment (Luwangel) [reply src/ui.py:88]: It avoids a variable. But it's not a big deal :).

File: src/Game.py [hunk 7]
```diff
@@ -32,4 +35,82 @@ def initGame(config1, config2, size):
...
+
+    if attacker == PLAYER_1:
+        boardsCopy[PLAYER_2]["life"] -= 1
+    else:
+        boardsCopy[PLAYER_1]["life"] -= 1
+
+    return boardsCopy
+
+def gameIsWon(boards, player):
+    gameIsWon = True
```
Comment (HALLERPierre) [inline src/Game.py:51]: No need to use a var. `return false` where you set GameIsWon to `false` and `return true` after the loops

File: src/Game.py [hunk 8]
```diff
@@ -32,4 +35,82 @@ def initGame(config1, config2, size):
...
+        updatedGameState, hit = updateBoardsAndHit(coord, gameState, PLAYER_1, PLAYER_2)
+    else:
+        updatedGameState, hit = updateBoardsAndHit(coord, gameState, PLAYER_2, PLAYER_1)
+
+    shipSunk = False
+
+    if hit:
+        # check entire ship:
+        # 1 find ship line in opponent ships
+        if currentPlayer == PLAYER_1:
```
Comment (HALLERPierre) [inline src/Game.py:80]: No need for condition :
`ships = gameState[getOpponent(currentPlayer)]["ships"]`

File: src/Game.py [hunk 9]
```diff
@@ -32,4 +35,82 @@ def initGame(config1, config2, size):
...
+
+        shipLine = []
+
+        for ship in ships:
+            for partCoord in ship:
+                if partCoord == coord:
+                    shipLine = ship
+                    break
+
+        # 2 check each coord of the ship with isShipPartHit
```
Comment (HALLERPierre) [inline src/Game.py:93]: I don't understand this part

File: src/Game.py [hunk 10]
```diff
@@ -32,4 +35,82 @@ def initGame(config1, config2, size):
...
+        shipSunk = True
+
+        for i in shipLine:
+            x, y = getCoords(i)
+            if updatedGameState[currentPlayer]["opponent"][x][y] != SQUARE_SUCCESS_SHOT:
```
Comment (HALLERPierre) [inline src/Game.py:98]: I am not sure what `opponent` is in `currentPlayer` gameState. Maybe rename to `opponent_board`  or `['opponent']['board']` ?

**Review by HALLERPierre — CHANGES_REQUESTED** [review]

File: src/Game.py (hunk 9 above)
Comment (benjaminpayoux) [reply src/Game.py:93]: After having found the ship to verify, I need to look into the opponent board and test if all the coordinates of the ship are hit. This tells me if the ship has sunk.

File: src/Game.py (hunk 10 above)
Comment (benjaminpayoux) [reply src/Game.py:98]: Yes, this is the opponent board. Will rename

File: src/ui.py [hunk 11]
```diff
@@ -1,33 +1,127 @@
...
-    if (currentPlayer == PLAYER_1):
-        displayBoard(boards[PLAYER_1]["opponent"])
+    if (player == PLAYER_1):
+        displayPlayerFleet(boards, PLAYER_1)
+        print()
+        displayBoard(boards[PLAYER_1]["opponent"], "PLAYER 1'S TURN")
+    elif player == PLAYER_2:
+        displayPlayerFleet(boards, PLAYER_2)
+        print()
+        displayBoard(boards[PLAYER_2]["opponent"], "PLAYER 2'S TURN")
```
[the remark this answers is not in the document]
Comment (benjaminpayoux) [reply src/ui.py:33]: :+1:

File: src/Game.py (hunk 4 above)
Comment (benjaminpayoux) [reply src/Game.py]: As the conception evolve, I think this will be delete in a future PR

**Pull request merged**